lib/utils/errors.go

```diff
@@ -86,6 +89,70 @@ func IsUntrustedCertErr(err error) bool {
 		strings.Contains(errMsg, "certificate is not trusted")
 }
 
+// CanExplainNetworkError returns a simple to understand error message that can
+// be used to debug common network and/or protocol errors.
+func CanExplainNetworkError(err error) (string, bool) {
+	var oerr *net.OpError
+	var derr *net.DNSError
+
+	switch {
+	// Connection refused errors can be reproduced by attempting to connect to a
+	// host:port that no process is listening on. The raw error typically looks
+	// like the following:
+	//
+	// dial tcp 127.0.0.1:8000: connect: connection refused
+	case errors.Is(err, syscall.ECONNREFUSED):
+		return "Connection refused. Run \"nc -vz a.b.c.d PORT\" on the Teleport " +
+			"agent to verify the target application is running and listening on " +
+			"the expected host and port.", true
+	// Host unreachable errors can be reproduced by running
+	// "ip route add unreachable a.b.c.d" to update the routing table to make
+	// the host unreachable. Packets will be discarded and an ICMP message
+	// will be returned. The raw error typically looks like the following:
+	//
+	// dial tcp 10.10.10.10:8000: connect: no route to host
+	case errors.Is(err, syscall.EHOSTUNREACH):
+		return "No route to host. Run \"ip route get a.b.c.d\" on the Teleport " +
+			"agent to verify a route to the target application exists in the " +
+			"routing table.", true
+	// Connection reset errors can be reproduced by creating a HTTP server that
+	// accepts requests but closes the connection before writing a response. The
+	// raw error typically looks like the following:
+	//
+	// read tcp 127.0.0.1:49764->127.0.0.1:8000: read: connection reset by peer
+	case errors.Is(err, syscall.ECONNRESET):
+		return "Connection reset by peer. Run \"curl -v a.b.c.d\" on the Teleport " +
+			"agent to verify the target application (or a load balancer in the " +
+			"network path) is not abruptly closing the connection after accepting it.", true
+	// I/O timeouts can be reproduced by creating a server with a customer
+	// listener that will time.Sleep after Accept(). The raw error typically
+	// looks like the following:
+	//
+	// dial tcp 127.0.0.1:8000: i/o timeout
+	case errors.As(err, &oerr) && oerr.Timeout():
+		return fmt.Sprintf("Network I/O timeout. Run \"nc -vz a.b.c.d\" on the "+
+			"Teleport agent to verify the target application is accepting network "+
+			"connections in under %v.", defaults.DefaultDialTimeout), true
+	// Slow responses can be reprodued by creating a HTTP server that does a
+	// time.Sleep before responding. The raw error typically looks like the following:
+	//
+	// context deadline exceeded
+	case errors.Is(err, context.DeadlineExceeded):
+		return fmt.Sprintf("Timeout waiting for response. Run \"curl -v a.b.c.d\" on the "+
+			"Teleport agent to verify the target application is responding to "+
```

Comment:
Similarly, if we want these errors to be specific to application access, then I would move this code into `lib/srv/app` or somewhere similar.